### PR TITLE
Accurately reflect pairing status on cached levels

### PR DESCRIPTION
**Pull request by jamescodeorg**
This change reworks and extends the changes in PR #43716 so that the user menu on the levels page for cached levels is populated from the existing `/api/user_menu` endpoint.  This fixes the issue where cached levels did not properly show pairing status.

This is intended to be a short-term fix to support this scenario and will be revisited when we separate pegasus from dashboard.  This fix is not an ideal long-term fix because of the following:
- The `/api/user_menu` endpoint returns a rendered user_menu in html.  Ideally, we would probably have an api return the data needed, put that data in redux and have react components that connected to that data.  This would require rewriting the existing haml-based rendering of the user menu to react.  There are also some additional complexities related to how this code is shared between pegasus and dashboard which will hopefully be clarified as we separate pegasus and dashboard.
- The `/api/user_menu` endpoint currently returns both the create menu and the sign-in button / user menu.  At the point where we call the api, we don't have the level information needed to determine whether the create menu should be shown.  As a hacky-ish workaround, we plumb through a flag that allows this code path to maintain the current state of the create menu.  A better solution would be to separate the create menu from the sign-in button / user menu but that was a bit more complicated, again due to how pegasus and dashboard share code here.

https://user-images.githubusercontent.com/71532552/150903785-552b080b-0b3d-460e-93f1-65fcbbf29a15.mov

## Links

- jira ticket: [LP-2132](https://codedotorg.atlassian.net/browse/LP-2132)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: shared/haml/user_header.haml [hunk 1]
```diff
@@ -2,23 +2,33 @@
 :ruby
   require 'cdo/create_header'
   require 'cdo/language_dir'
-  limit_project_types = request.cookies["_limit_project_types_#{rack_env}"]
 
-  level = defined?(local_assigns) ?
-    local_assigns[:level] :
-    false
+  # Determine if we want to show the create menu
+  if local_assigns[:show_create_menu]
```
Comment (maureensturgeon) [inline shared/haml/user_header.haml:7]: Do you need to check `defined?` here? (`defined?(local_assigns)`)

File: shared/haml/user_header.haml [hunk 2]
```diff
@@ -2,23 +2,33 @@
...
-    project_type: params && params[:key]
-  }
-  create_drop_down_items = CreateHeader.get_create_dropdown_contents(options)
+  if show_create_menu
+    limit_project_types = request.cookies["_limit_project_types_#{rack_env}"]
 
-  # We want to show the Create button for all pages except when working on a
-  # course level. Specifically, for any page except non-project levels.
-  should_show_create_menu = !level || level.try(:is_project_level)
+    level = defined?(local_assigns) ?
```
Comment (maureensturgeon) [inline shared/haml/user_header.haml]: Do you need to keep this? Is `level` used anywhere?

File: dashboard/app/views/layouts/_header.html.haml [hunk 3]
```diff
@@ -156,3 +156,9 @@
 - elsif level
   :javascript
     dashboard.header.buildProjectInfoOnly()
+
+-# If the page is cached, the sign-in button / user menu must be loaded separately
+-# by the client.
+- if @public_caching
+  :javascript
+    dashboard.header.buildUserMenu()
```
Comment (maureensturgeon) [inline dashboard/app/views/layouts/_header.html.haml:164]: I forget, is this also called in pegasus? For pegasus cached pages is the pairing status right?

**Review by maureensturgeon — APPROVED** [review]
Thanks for the video in the description! I just have a couple small comments/questions

File: dashboard/app/views/layouts/_header.html.haml (hunk 3 above)
Comment (jamescodeorg) [reply dashboard/app/views/layouts/_header.html.haml:164]: No, this is dashboard only.  The pairing status for pegasus may still be wrong.  (The main goal of this fix was to fix this issue for the levels and script overview pages, but it does fix it for all of dashboard.)

File: shared/haml/user_header.haml (hunk 2 above)
Comment (jamescodeorg) [reply shared/haml/user_header.haml]: No, I don't need this anymore, nice catch!

File: shared/haml/user_header.haml (hunk 1 above)
Comment (jamescodeorg) [reply shared/haml/user_header.haml:7]: My understanding is that `local_assigns` is a method (https://apidock.com/rails/v5.0.0.1/ActionView/Template/local_assigns) so we don't need to check if it's `defined?`  (I think the original code wasn't checking what it looked like it was checking for.)

**Review by maureensturgeon — APPROVED** [review]

**Pull request merged**